zulip/integrations/bridge_with_matrix/matrix_bridge.py

```diff
@@ -33,146 +49,352 @@ class Bridge_FatalMatrixException(Exception):
-    except MatrixRequestError as exception:
-        if exception.code == 403:
-            raise Bridge_FatalMatrixException("Bad username or password.")
-        else:
-            raise Bridge_FatalMatrixException("Check if your server details are correct.")
-    except MissingSchema:
-        raise Bridge_FatalMatrixException("Bad URL format.")
+class MatrixToZulip:
+    """
+    Matrix -> Zulip
+    """
 
+    non_formatted_messages = {
+        StickerEvent: "sticker",
+    }
+
+    def __init__(
+        self,
+        matrix_client: AsyncClient,
+        matrix_config: Dict[str, Any],
+        zulip_client: zulip.Client,
+        no_noise: bool,
+    ) -> None:
+        self.matrix_client = matrix_client
+        self.matrix_config = matrix_config
+        self.zulip_client = zulip_client
+        self.no_noise = no_noise
+
+    @classmethod
+    async def create(
+        cls,
+        matrix_client: AsyncClient,
+        matrix_config: Dict[str, Any],
+        zulip_client: zulip.Client,
+        no_noise: bool,
+    ) -> "MatrixToZulip":
+        matrix_to_zulip = cls(matrix_client, matrix_config, zulip_client, no_noise)
+
+        # Login to Matrix
+        await matrix_to_zulip.matrix_login()
+        await matrix_to_zulip.matrix_join_rooms()
+
+        # Do an initial sync of the matrix client in order to continue with
+        # the new messages and not with all the old ones.
+        sync_response = await matrix_client.sync()
+        if isinstance(sync_response, SyncError):
+            raise Bridge_FatalMatrixException(sync_response.message)
+
+        return matrix_to_zulip
+
+    async def _matrix_to_zulip(self, room: MatrixRoom, event: Event) -> None:
+        # We do this to identify the messages generated from Zulip -> Matrix
+        # and we make sure we don't forward it again to the Zulip stream.
+        if event.sender == self.matrix_config["mxid"]:
+            return
 
-def matrix_join_room(matrix_client: Any, matrix_config: Dict[str, Any]) -> Any:
-    try:
-        room = matrix_client.join_room(matrix_config["room_id"])
-        return room
-    except MatrixRequestError as exception:
-        if exception.code == 403:
-            raise Bridge_FatalMatrixException("Room ID/Alias in the wrong format")
-        else:
-            raise Bridge_FatalMatrixException("Couldn't find room.")
+        if room.room_id not in self.matrix_config["bridges"]:
+            return
+        stream, topic = self.matrix_config["bridges"][room.room_id]
 
+        content = await self.get_message_content_from_event(event, room)
+        if not content:
+            return
 
-def die(signal: int, frame: FrameType) -> None:
-    # We actually want to exit, so run os._exit (so as not to be caught and restarted)
-    os._exit(1)
+        # Update the bot's read marker in order to show the other users which
+        # messages are already processed by the bot.
+        await self.matrix_client.room_read_markers(
```

Comment:
I think the mark-as-read step should only be done once the relay to Zulip is successful. So that if it fails, the bot can retry again.

Reply:
absolutely, thanks!